pkg/image/index/index.go

```diff
@@ -0,0 +1,179 @@
+		},
+		cache.Indexers{
+			ChainIndexName:     chainIndexFunc,
+			LastChainIndexName: lastChainIndexFunc,
+		},
+	)
+
+	go c.imageController.Run(stopChan)
+	return c
+}
+
+func (c *ImageIndex) WaitForSyncedStores() {
+	for !c.imageController.HasSynced() {
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+// Descendants provides a list of images that the provided image is based on.
+// Note that if a base image is tagged multiple times, this will return all the matching
+// images.
+func (c *ImageIndex) Descendants(image *imageapi.Image) ([]*imageapi.Image, error) {
+	imageChains := getImageLayerChain(image)
+	result := &uniqueImageList{}
+	for _, chain := range imageChains {
+		objects, err := c.imageIndex.ByIndex(LastChainIndexName, chain)
+		if err != nil {
+			return nil, err
+		}
+		for _, obj := range objects {
+			if i := obj.(*imageapi.Image); i.Name != image.Name {
+				result.Append(i)
+			}
+		}
+	}
+	return result.Images, nil
+}
+
+// Ancestors provides a list of images that are based in the provideed image.
+func (c *ImageIndex) Ancestors(image *imageapi.Image) ([]*imageapi.Image, error) {
+	chains := getImageLayerChain(image)
+	objects, err := c.imageIndex.ByIndex(ChainIndexName, chains[len(chains)-1])
+	if err != nil {
+		return nil, err
+	}
+	result := &uniqueImageList{}
+	for _, obj := range objects {
+		i := obj.(*imageapi.Image)
+		imageChain := getImageLayerChain(i)
+		// Remove the image itself from the result
+		if imageChain[len(imageChain)-1] != chains[len(chains)-1] {
+			result.Append(i)
+		}
+	}
+	return result.Images, nil
+}
+
+func (c *ImageIndex) addImage(obj interface{}) {
+	if err := c.imageIndex.Add(obj); err != nil {
+		glog.Errorf("failed to add image %+v to index: %v", obj, err)
+	}
+}
+
+func (c *ImageIndex) updateImage(_, newObj interface{}) {
+	if err := c.imageIndex.Update(newObj); err != nil {
+		glog.Errorf("failed to update image %+v in index: %v", newObj, err)
+	}
+}
+
+func (c *ImageIndex) deleteImage(obj interface{}) {
+	i, ok := obj.(*imageapi.Image)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			glog.Errorf("couldn't get object from tombstone %#v", obj)
+			return
+		}
+		i, ok = tombstone.Obj.(*imageapi.Image)
+		if !ok {
+			glog.Errorf("tombstone contained object that is not a Deployment %#v", obj)
+			return
+		}
+	}
+	c.imageIndex.Delete(i)
+}
+
+type uniqueImageList struct {
+	Images []*imageapi.Image
+	index  map[string]struct{}
+}
+
+func (l *uniqueImageList) Append(image *imageapi.Image) {
+	if l.index == nil {
+		l.index = map[string]struct{}{}
+	}
+	if _, exists := l.index[image.Name]; exists {
+		return
+	}
+	l.Images = append(l.Images, image)
+	l.index[image.Name] = struct{}{}
+}
+
+// getImageLayerChain recalculates the digest from each layer and return a sequence that
+// represents the image.
+func getImageLayerChain(image *imageapi.Image) []string {
+	chain := []string{}
+	for i, layer := range image.DockerImageLayers {
```

Comment:
@miminar The image is the totality of all layers in a specific order. These chains are not about content but about the restoration of the inheritance because the hash of each layer are not unique. Metadata is an important part of the image.